File: Mariani_Giada_SCALA/src/tris/Tris.scala
/**
 * @author Mariani Giada
 * 
 * Progetto Info 3 - A.
 * Gioco del tris.
 * 
 */

package tris

/*
 * Importazione 
 */
import scala.collection.mutable.HashMap

/*
 * TRIS
 */
object Tris {

  /*
   * MAIN 
   */
  def main(args: Array[String]): Unit = {

    //DICHIARAZIONE VARIABILI
    val utenti = List ("UTENTE1", "UTENTE2")
    val coordinate = List(1, 2, 3, 4, 5, 6, 7, 8, 9)
    val vincitore = List("UTENTE1", "UTENTE2", "PATTA")
    var winner = "PATTA"
    var fineGioco = false
    var board: Array[String] = new Array[String](9) 
    var turno = 1
    
    //STAMPA TURNO
    def stampaTurno(t: Int){
      if(t == 1) println("Turno utente 1")
      else
        println("Turno utente 2")
    }
    
    //CAMBIO DEL TURNO
    def cambiaTurno(t:Int){
      if(t == 1) turno = 2
      else turno = 1
    }
    
    //SCELTA COORDINATE UTENTE 1 
    def sceltaUtente1(){
      var c = 1
      var i = false
      do{
        println("U1: Inserisci casella da scegliere \n(0-1-2\n3-4-5\n6-7-8)\n")
        c = readInt
        if(c >= 0 && c <= 8){
          if(board(c) == null){
            board(c) = "UTENTE1"
            i = true
          }
        }
      }while((c < 0 || c > 8) || i == false)
    }
    
    //SCELTA COORDINATE UTENTE 2
    def sceltaUtente2(){
      var c = 1
      var i = false
      do{
        println("U2: Inserisci casella da scegliere \n(0-1-2\n3-4-5\n6-7-8)\n")
        c = readInt
        if(c >= 0 && c <= 8){
          if(board(c) == null){
            board(c) = "UTENTE2"
            i = true
          }
        }
      }while((c < 0 || c > 8) || i == false)
    }

    //FUNZIONE VINCITA
    def vincita(){
      if(
          (
              ((board(0) == board(1) && board(1) == board(2)) && (board(0) == "UTENTE1" || board(0) == "UTENTE2"))
          || ((board(3) == board(4) && board(4) == board(5)) && (board(3) == "UTENTE1" || board(3) == "UTENTE2"))
          || ((board(6) == board(7) && board(7) == board(8)) && (board(6) == "UTENTE1" || board(6) == "UTENTE2"))
          || ((board(0) == board(3) && board(3) == board(6)) && (board(0) == "UTENTE1" || board(0) == "UTENTE2"))
          || ((board(1) == board(4) && board(4) == board(7)) && (board(1) == "UTENTE1" || board(1) == "UTENTE2"))
          || ((board(2) == board(5) && board(5) == board(8)) && (board(2) == "UTENTE1" || board(2) == "UTENTE2"))
          || ((board(0) == board(4) && board(4) == board(8)) && (board(0) == "UTENTE1" || board(0) == "UTENTE2"))
          || ((board(2) == board(4) && board(4) == board(6)) && (board(2) == "UTENTE1" || board(2) == "UTENTE2"))
          )
         ){
        fineGioco = true
      }
    }

    //FUNZIONE CASELLE PIENE - PATTA
    def casellePiene(){
      if(board(0) == null || board(1) == null || board(2) == null ||  
         board(3) == null || board(4) == null || board(5) == null ||
         board(6) == null || board(7) == null || board(8) == null )
        fineGioco = false
      else
        fineGioco = true
    }
    
    //STAMPA BOARD
    def stampaBoard(){
      var i = 0
      
      for(i <- 0 to 2){
         print(board(i) + " | ")
      }
      
      println()
      
      for(i <- 3 to 5){
         print(board(i) + " | ")
      }
      
      println()
      
      for(i <- 6 to 8){
         print(board(i) + " | ")
      }
      
      println()
      println()
    }

    
    //FUNZIONAMENTO DEL GIOCO
    board.foreach { x => null }
    
    while(fineGioco == false){
      
      stampaBoard()
      stampaTurno(turno)
      
      if(turno == 1){
        sceltaUtente1()
        vincita()
        if(fineGioco == true)
          winner = "UTENTE1"
        cambiaTurno(turno)
      }
      else{
        sceltaUtente2()
        vincita()
        if(fineGioco == true)
          winner = "UTENTE2"
        cambiaTurno(turno)
      }
      
      casellePiene()
    }
    
    println("Vincitore: " + winner)
  }
}

File: Mariani_Giada_C++/src/Tris.cpp
/*
 * Tris.cpp
 *
 *  Created on: 30 dic 2015
 *      Author: Giada
 */

#include <iostream>
#include <string.h>
using namespace std;

class Tris{
private:
	int board[3][3];
	int turno;
	int winner;
	bool fineGioco;

public:

	Tris(){
		int i,j;
		for(i=0;i<3;i++){
			for(j=0;j<3;j++){
				board[i][j] = 0;
			}
		}

		winner = 0;
		fineGioco = false;
		turno = 1;
	}

	void cambiaTurno(){
		if(turno == 1)
			turno = 2;
		else
			turno = 1;
	}

	int stampaTurno(){
		return turno;
	}

	void setBoard(int u){
		int r, c;
	    bool i = false;
	      do{
	        cout<<"Inserisci riga: "<<endl;
	        cin>>r;
	        cout<<"Inserisci colonna: "<<endl;
	        cin>>c;

	        if((r >= 0 && r <= 2) || (c>=0 && c<=2)){
	          if(board[r][c] == 0){
	            board[r][c] = u;
	            i = true;
	          }
	        }
	      }while((r<0 || r>2 || c<0 || c>2) || i == false);
	}

	void stampaBoard(){
		int r, c;
		for(r = 0; r < 3; r++){
			for(c = 0; c < 3; c++){
				cout<< board[r][c]<< " | ";
			}
			cout<<endl;
		}
	}

	void vincita(){
	      if((((board[0][0] == board[0][1] && board[0][1] == board[0][2]) && (board[0][0] == 1 || board[0][0] == 2))
	          || ((board[1][0] == board[1][1] && board[1][1] == board[1][2]) && (board[1][0] == 1 || board[1][0] == 2))
	          || ((board[2][0] == board[2][1] && board[2][1] == board[2][2]) && (board[2][0] == 1 || board[2][0] == 2))
	          || ((board[0][0] == board[1][0] && board[1][0] == board[2][0]) && (board[0][0] == 1 || board[0][0] == 2))
	          || ((board[0][1] == board[1][1] && board[1][1] == board[2][1]) && (board[0][1] == 1 || board[0][1] == 2))
	          || ((board[0][2] == board[1][2] && board[1][2] == board[2][2]) && (board[0][2] == 1 || board[0][2] == 2))
	          || ((board[0][0] == board[1][1] && board[1][1] == board[2][2]) && (board[0][0] == 1 || board[0][0] == 2))
	          || ((board[0][2] == board[1][1] && board[1][1] == board[2][0]) && (board[0][2] == 1 || board[0][2] == 2))
	          )
	         )
	        fineGioco = true;
	}

	void casellePiene(){
		if(!(board[0][0] == 0 || board[0][1] == 0 || board[0][2] == 0
		|| board[1][0] == 0 || board[1][1] == 0 || board[1][2] == 0
		|| board[2][0] == 0 || board[2][1] == 0 || board[2][2] == 0))
			fineGioco = true;
	}

	bool fineDelGioco(){
		return fineGioco;
	}

	void setWinner(int w){
		winner = w;
	}

	int vincitore(){
		return winner;
	}
};


int main(){

	Tris tris;

	while(tris.fineDelGioco() != true){
		if(tris.stampaTurno() == 1){
			cout<<endl<<"U1:"<<endl;
			tris.setBoard(1);

			tris.vincita();
			if(tris.fineDelGioco() == true){
				tris.setWinner(1);
			}

			tris.cambiaTurno();
		}
		else{
			cout<<endl<<"U2:"<<endl;
			tris.setBoard(2);

			tris.vincita();
			if(tris.fineDelGioco() == true){
				tris.setWinner(2);
			}

			tris.cambiaTurno();
		}

		tris.stampaBoard();
		tris.casellePiene();
	}

	if(tris.vincitore() == 1)
		cout<<"Vincitore: UTENTE1"<<endl;
	if(tris.vincitore() == 2)
		cout<<"Vincitore: UTENTE2"<<endl;
	if(tris.vincitore() == 0)
		cout<<"Vincitore: PATTA"<<endl;
}
